### PR TITLE
Update has_block_template function to use apply_filters

### LINKED ISSUE #4970: Add wc_has_block_template filter to WooCommerce Core
## Description
As described in https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/4939 we need to update the return value of the following [function](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/includes/class-wc-template-loader.php#L110-L112)

From
```php
return is_readable(
    get_stylesheet_directory() . '/block-templates/' . $template_name . '.html'
);
```

To
```php
return apply_filters( 'wc_has_block_template',
    is_readable(
        get_stylesheet_directory() . '/block-templates/' . $template_name . '.html'
    )
);
```

This will allow us to override this value from Woo Blocks to serve the block templates from Woo Blocks.

## Acceptance Criteria
This will just require some regression testing to make sure everything is working as expected still.

* Given the same scenarios of rendering Block Templates over Legacy PHP Templates if they're available within the theme of a compatible Block Theme, everything should still work the same.

### PR DESCRIPTION
**Pull request by tjcafferkey**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Adding an `apply_filters( 'woocommerce_has_block_template ')` will allow us to a step forward to rendering block templates from Woo Blocks in an effort to replace core PHP Templates.

Context found [here](https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/4926)

Closes https://github.com/woocommerce/woocommerce-gutenberg-products-block/issues/4970

### How to test the changes in this Pull Request:

This will need some regression testing to confirm we haven't introduced any adverse side effects.

1. Activate a block template enabled theme.
2. Add `single-product.html` in `theme-dir/block-templates` with the following content.

```html
<!-- wp:template-part {"slug":"header","tagName":"header","className":"site-header","layout":{"inherit":true} /-->
<!-- wp:paragraph --> <p>You are using the block template now</p> <!-- /wp:paragraph -->
<!-- wp:template-part {"slug":"footer","tagName":"footer","className":"site-footer","layout":{"inherit":true} /-->
```

3. Check on the frontend product page that this template is getting served to the user instead of the core PHP template.
4. Delete the `single-product.html` template file and confirm that the core PHP template is getting rendered
5. Activate a classic theme and confirm core PHP template is getting rendered there.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?

### Changelog entry

> Update has_block_template method: Add apply_filters to the function which will enable third-party plugins to override the return value.

### FOR PR REVIEWER ONLY:

* [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: plugins/woocommerce/includes/class-wc-template-loader.php
Comment (Aljullu) [inline plugins/woocommerce/includes/class-wc-template-loader.php]: Question for the WC core reviewer: is there any preference on using `wc_` or `woocommerce_` as the filter prefix? I noticed there are filters with both prefixes, but `woocommerce_` seems to be more popular?

**Review by Aljullu — DISMISSED** [review]
Approving from the WC Blocks side.

This will still need a review from a maintainer of WC core, though.

File: plugins/woocommerce/includes/class-wc-template-loader.php
Comment (barryhughes) [reply plugins/woocommerce/includes/class-wc-template-loader.php]: That's right! `woocommerce_` as prefix [is preferred](https://github.com/woocommerce/woocommerce/wiki/Naming-conventions#php) and is what new code should use. Some other thoughts on this hook:

- An additional param containing the `$template_name` may be useful (otherwise callbacks will not necessarily know which template is being assessed).
- Casting the result `(bool)` could add an extra layer of safety (to ensure rogue callbacks don't change type on us).
- A docblock may be useful.

**Review by barryhughes — CHANGES_REQUESTED** [review]
Nice, works well. Left a few notes/requests re the new hook.

File: plugins/woocommerce/includes/class-wc-template-loader.php
Comment (barryhughes) [inline plugins/woocommerce/includes/class-wc-template-loader.php]: Hi! My earlier comment probably wasn't clear enough, but this is what I meant re applying a cast (basically, apply it to the `apply_filters()` call before we return):

```suggestion
		$has_template = is_readable( get_stylesheet_directory() . '/block-templates/' . $template_name . '.html' );

		/**
		 * Filters the value of the result of the block template check.
		 *
		 * @since x.x.x
		 *
		 * @param boolean $has_template value to be filtered.
		 * @param string $template_name The name of the template.
		 */
		return (bool) apply_filters( 'woocommerce_has_block_template', $has_template, $template_name );
```

To explain: the [`is_readable()`](https://www.php.net/manual/en/function.is-readable.php) function always returns a boolean, so no cast needed there. However, this filter can be used by anyone, including 3PDs, and they may unintentionally change the type during filtering. So, the cast is here to guarantee the declared return type per the method's docblock.

**Review by barryhughes — CHANGES_REQUESTED** [review]
Should relocate the cast to boolean.

File: plugins/woocommerce/includes/class-wc-template-loader.php
Comment (tjcafferkey) [reply plugins/woocommerce/includes/class-wc-template-loader.php]: Ah apologies @barryhughes . Makes total sense, not sure how I misinterpreted that looking back. I've committed you suggestion. Thank you.

**Review by barryhughes — APPROVED** [review]
LGTM!

**Comment by barryhughes** [comment]
One of the tests failed, probably a temp issue/unrelated glitch—but going to re-run before merging just to be safe.

**Pull request merged**
